### ISSUE TITLE
PLDS problems seem to be due to misaligned input traces

### ISSUE THREAD

**Issue by mitrajz** [joacorapela/ldsForNeuralPopulation#9]
I've been looking into the problem with the PLDS models, especially the fact that the sates don't change much with the visual inputs, while we know V1 neurons are very strongly driven by the visual inputs. Here is an example of this problem for one animal. Green patch denotes the presence of visual input, and the black lines are the 9 states aligned to the stimulus onset and averaged over multiple occurrences of the same input ( = the PSTH of each state) 
![lag0_states](https://user-images.githubusercontent.com/40278271/125981797-eb9d047e-f54c-49db-8bfe-1662a95b3993.png)
I noticed that something is off when plotting the average firing rate across all V1 neurons in the same time window:
![lag0_fr](https://user-images.githubusercontent.com/40278271/125982269-c4ed65ec-947a-4922-b598-3e30603153fe.png)
The firing rate increases before the visual input. The model can't capture this, as ut can only influence xt, and xt+k (k>0) through the dynamics, and not xt-1. I am not sure how you've done the binning, but something gets misaligned there or the binning is not causal, as this is obviously not the case with the raw data. I did a quick fix by pushing the input traces one time step back, leading to a causal response from the neurons:
![lag1_fr](https://user-images.githubusercontent.com/40278271/125982807-e4c18c82-bc29-467b-bca4-74fa682e7559.png)
 Using these traces to fit the model gives these states:
![lag1_states](https://user-images.githubusercontent.com/40278271/125982850-f98f392d-aa7a-47b5-a909-da5f31af6381.png)
Which is a completely different picture and now all the states are changing with the onset of the visual input.
The model performance also improves. Here, the variational lower bound is plotted against the EM iterations, which you can see reaches higher values when the input lag is fixed (solid blue line), and in addition, the convergence is also much faster in this case.
![Varbounds](https://user-images.githubusercontent.com/40278271/125983395-e1165dc3-1d78-4b74-9c4d-5d3e0c6e5d11.png)
So, the problem seems to be with the binning and not fitting the models.

**Comment by joacorapela** [comment]
I will try to take a look at this during the weekend Mitra.

**Comment by joacorapela** [comment]
I did a quick check of comparing the spike counts and inputs I used in the R and Matlab estimations. From these quick tests it seems that they are both identical. So the binning problem may not explain the differences between the GPLDS and the PLDS.

However, your second plot is alarming and it may be that all stimuli are advanced one sample in time. Fixing this problem may improve the performance of both models.

Please note that you cannot compare the variational lower bound between estimates using different stimuli.However the increased modulations that you showed in the figure before the last one seem to indicate that there is a problem in the alignment of inputs.

Could it be that the alignment problem appears in the raw data? I will look more into this on Wednesday.

**Comment by joacorapela** [comment]
The binning of the spike times and inputs is done in https://github.com/joacorapela/ldsForNeuralPopulation/blob/master/matlabCode/scripts/doSaveBinnedLDSTimeSeries.m

I will double check it on Wed, but it seems straight forward.

**Comment by mitrajz** [comment]
No, the raw data is definitely correct, you can check in the original .mat files.
Re GLDS vs. PLDS, this could be due to the fact that GLDS has an additional channel for the inputs to influence the states, though the d inputs and D input weights. d inputs modulate the observations directly and not through the states, so the way their influence decays over time is not strictly an exponential decay (as is the case with the u inputs), and they can capture any temporal pattern in the input influence through an arbitrary length for the memory. This could mean that adding d inputs makes the model less sensitive to any temporal misalignment of inputs with neural recordings.

**Comment by mitrajz** [comment]
So, I looked into how binning is done and here is what the problem is: Let's say there is a bin specified with 2 edges, edge_1 and edge_2. In the current code, when you bin the stimulus, if the visual stim goes on inside this bin, the binned value will be 0 for edge_1 and 1 for edge_2. However, for spike, you use histcounts that does the opposite, if a spike is between edge_1 and edge_2, it is included in edge_1. So, in a way, the stimulus is rounded up in time while the spike are rounded down in time. That's why they are misaligned.

**Comment by mitrajz** [comment]
In general, one needs to be careful with binning  the stimulus times, as their onset does not necessarily fall into our bins and assigning it to either the lower or higher bin is not really accurate. Instead, the way these things are usually done is to align the onset of the bins to the stimulus onset directly. This is only doable if the analysis is per trial, in which case the stimulus onset for each trial is time 0 and the onset of a bin, and then we bin backwards and forwards in time ( = PSTH).  For the continuous analysis, one option is to change the size of one bin between each 2 trials slightly, to make sure the stimulus onsets are all on the bin edges.

**Comment by joacorapela** [comment]
This is a tricky issue Mitra. The key question is if we are going to say that stimulus was on in a time bin when it was on only on a small fraction of the bin duration. For example, suppose that a bin starts at time 0ms and ends at time 100ms, and that stimulation starts at tim 99.9ms and ends at time 149.99ms. Shall we say that stimulus was on on this bin, when it was on for only 1% of the bin duration?

In the code I chose to say that a stimulus was on on a bin only when it was completely stimulated. With this choice we can get a burst of spikes after the onset of the stimulus in a time bin labelled as non-stimulated.

I can chose to say that a stimulus was on on a partially stimulated bin. However in this case we could see that the spike rate on stimulated bin is low due to the small number of spikes happening before the onset of a stimulus occurring very close to the end of the bin. The pull request that you submitted could cause this problem.

Considering the delay between when a stimulus is presented and when the information reaches V1 or LM, I think that my choice of labelling a partially stimulated time bin as non-stimulated makes sense. 

What do you think?

**Comment by joacorapela** [comment]
Also, note that this binning problem does not happen with spike times, which are unambiguously assigned to one bin.

**Comment by mitrajz** [comment]
This does not address my point. The issue is not the interpretation of binning and whether the stimulus onset should be assigned to the bin before or after. The issue is that in your code, binning is handled differently for spikes vs the stimulus trace. You are assigning the spikes to edge before the bin (using histcounts), while assigning the stimulus to the edge after the bin. This is incorrect and is exactly the reason that we see neurons spiking before the stimulus arrives in your binned data.
The correct way to tackle this, is to align everything directly to the stimulus onset, as I explained before and is the way it is traditionally done. However, before implementing this, for the moment, the very least thing we should do is to ensure that spikes and stimulus are binned with a consistent logic, either both to the bin before or both to the bin after. If you would like to continue using histcount that assigns spikes to the bin before, then the stimulus should also be assigned to the previous bin.
Re the lag between the stimulus presentation and spikes, you should not add your assumptions of biology into the way you bin the data.

**Comment by joacorapela** [comment]
It seems you are not understanding my comments Mitra. Spikes are assigned unambiguously to their correct bin. But determining the first stimulated bin is ambiguous and I argue that the choice I chose is more convenient. If you have time let's talk about this tomorrow. I am not enjoying this discussion.

**Comment by mitrajz** [comment]
Let's talk tomorrow then, let me know what time you prefer.

**Comment by joacorapela** [comment]
Any time in the afternoon is good for me Mitra. Thanks.

**Comment by mitrajz** [comment]
Great, let's meet at 13:30 then, 4th floor east quad meeting room.
